replay.js

```diff
@@ -283,9 +283,42 @@ async function main() {
         }
     });
 
+    /** @type {import("typescript/lib/protocol").CompletionInfoResponse | undefined} */
+    let lastCompletionsResponse;
+    let lastCompletionEntryDetailsIndex = -1;
+
     for (const request of requests) {
         exitRequested = exitRequested || request.command === "exit";
+
         if (!unattended) console.log(`${request.seq}\t${request.command}`);
+
+        if (request.command === "completionEntryDetails") {
+            let detailsRequest = /** @type {import("typescript/lib/protocol").CompletionDetailsRequest} */ (request);
+            // completionEntryDetails can be very specific to the session and original platform,
+            // and apparently some information in `data` might differ between what the server expects vs. what the replay script contains.
+            for (const replayEntry of detailsRequest.arguments.entryNames) {
+                if (typeof replayEntry !== "string" && replayEntry.source !== undefined && typeof replayEntry.data?.fileName === "string") {
+                    if (!lastCompletionsResponse) {
+                        console.error("Cannot run 'completionEntryDetails' with 'source' and 'data' without running a preceding 'completionInfo'.");
+                        await exitOrKillServer();
+                        process.exit(5);
+                    }
+
+                    const actualEntries = lastCompletionsResponse.body.entries;
+                    // Typically these will be sequential, but we might need to loop back around.
+                    lastCompletionEntryDetailsIndex = findSuitableCompletionEntryDetailAndReplaceData(lastCompletionEntryDetailsIndex, actualEntries, replayEntry);
```

Comment:
I wanted to avoid going definitely-quadratic, but this double pass feels bad and a map on every completions request might start to add up?

Reply:
If everything else in the replay is accurate, the indices should always be the same, right? Also, is `detailsRequest.arguments.entryNames` ever longer than 1? AFAIK, nobody uses the API to get more than one set of details at a time.

Reply:
I'm just trying to be robust on that.

> If everything else in the replay is accurate, the indices should always be the same, right?

I thought so, but without the "loop around again" logic, it does fail on some repros. 😬

Similarly, we don't necessarily always find the same completion entry either!

Reply:
You could build a lookup table of `name`/`source`/`data` first, but it will probably be worse than what you’re doing now if `entryNames` is only 1 long.